src/v/config/bounded_property.h

```diff
@@ -0,0 +1,168 @@
+// Copyright 2022 Vectorized, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#pragma once
+
+#include "config/property.h"
+
+namespace config {
+
+/**
+ * Detail section for concepts used in public class definitions, avoid
+ * polluating the overall config:: namespace with these.
+ */
+namespace detail {
+
+/**
+ * Traits required for a type to be usable with `numeric_bounds`
+ */
+template<typename T>
+concept numeric = requires(const T& x) {
+    {x % x};
+    { x < x } -> std::same_as<bool>;
+    { x > x } -> std::same_as<bool>;
+};
+
+/**
+ * Concept that is true for stdlib containers which publish their
+ * inner contained type as ::value_type
+ */
+template<typename T>
+concept has_value_type = requires(T x) {
+    typename T::value_type;
+};
+
+/**
+ * inner_type is a struct whose ::inner member reflects
+ * the value_type of T if T has such an attribute.  Otherwise
+ * ::inner is equal to T.
+ *
+ * Useful when you expect either a std::optional<> or a bare value,
+ * but always want to use the bare value type.
+ */
+template<typename T>
+struct inner_type {
+    using inner = T;
+};
+
+template<typename T>
+requires has_value_type<T>
+struct inner_type<T> {
+    using inner = typename T::value_type;
+};
+
+} // namespace detail
+
+/**
+ * Define valid bounds for a numeric configuration property.
+ */
+template<typename T>
+requires detail::numeric<T>
+struct numeric_bounds {
+    std::optional<T> min = std::nullopt;
+    std::optional<T> max = std::nullopt;
+    std::optional<T> align = std::nullopt;
+
+    T clamp(T& original) {
+        T result = original;
+
+        if (align.has_value()) {
+            auto remainder = result % align.value();
+            result -= remainder;
+        }
+
+        if (min.has_value()) {
+            result = std::max(min.value(), result);
+        }
+
+        if (max.has_value()) {
+            result = std::min(max.value(), result);
+        }
```

Comment:
@jcsp should alignment be verified after min/max have been applied (or verify that min/max are also aligned when they change the result)?